File: euler-136-singleton-difference.go/euler-136-singleton-difference.go
package main

import (
	"fmt"
	"math"
	"os"
	"strconv"
)

// compute and count possible solutions for n = 1 to n = limit
func compute(limit int) int {
	// initialize an array to store the counts for each n
	array := make([]int, limit)

	// iterate over possible values of 'a'
	for a := 1; a < limit; a++ {
		// iterate over possible values of 'd'
		for d := int(math.Floor(float64(a)/4)) + 1; d < a; d++ {
			n := a * (4*d - a)

			// if n exceeds the limit, stop the loop for 'd'
			if n > limit-1 {
				break
			} else {
				// increment the count for the current value of n in the array
				array[n]++
			}
		}
	}

	// count how many times the solution for n was completely unique with only 1 array element
	count := 0
	for _, value := range array {
		if value == 1 {
			count++
		}
	}

	return count
}

func main() {
	// convert the command line argument to an integer
	var input int
	var err error
	input, err = strconv.Atoi(os.Args[1])
	if err != nil {
		// if there's an error in conversion, print it and exit
		panic(err)
	}
	result := compute(input)
	fmt.Printf("%d\n", result)
}

File: euler-136-singleton-difference.kt/euler-136-singleton-difference.kt
import kotlin.math.floor

// defining a solution counter
interface SolutionCounter {
    fun compute(): Int
}

// implements common functionality for solution counting
open class SolutionCounterBase(val limit: Int) : SolutionCounter {

    // stores the counts for each 'n'
    protected val array = IntArray(limit)

    // computes and counts possible solutions
    override fun compute(): Int {
        for (a in 1 until limit) {
            for (d in (floor(a.toDouble() / 4.0) + 1).toInt() until a) {
                val n = a * (4 * d - a)
                if (n > limit - 1) {
                    break
                } else {
                    array[n]++
                }
            }
        }

        // counts how many times the solution for 'n' was completely unique with only 1 array element.
        var count = 0
        for (value in array) {
            if (value == 1) {
                count++
            }
        }

        return count
    }
}


// Function demonstrating polymorphism by accepting any object that implements SolutionCounter interface
fun executeSolutionCounter(solutionCounter: SolutionCounter) {
    val result = solutionCounter.compute()
    println("$result")
}

fun main(args: Array<String>) {
	// Read the input from the command line
	val input = args[0].toInt()
    val solutionCounter: SolutionCounter = SolutionCounterBase(input)
    executeSolutionCounter(solutionCounter)
}
